resend_scheduler.go

```diff
@@ -1,148 +1,130 @@
 package nex
 
 import (
+	"context"
+	"fmt"
 	"time"
 )
 
-// TODO - REMOVE THIS ENTIRELY AND REPLACE IT WITH AN IMPLEMENTATION OF rdv::Timeout AND rdv::TimeoutManager AND USE MORE STREAM SETTINGS!
-
-// PendingPacket represends a packet scheduled to be resent
+// PendingPacket represents a packet scheduled to be resent
 type PendingPacket struct {
-	packet         PRUDPPacketInterface
-	lastSendTime   time.Time
-	resendCount    uint32
-	isAcknowledged bool
-	interval       time.Duration
-	ticker         *time.Ticker
-	rs             *ResendScheduler
-}
-
-func (pi *PendingPacket) startResendTimer() {
-	pi.lastSendTime = time.Now()
-	pi.ticker = time.NewTicker(pi.interval)
-
-	for range pi.ticker.C {
-		finished := false
-
-		if pi.isAcknowledged {
-			pi.ticker.Stop()
-			pi.rs.packets.Delete(pi.packet.SequenceID())
-			finished = true
-		} else {
-			finished = pi.rs.resendPacket(pi)
-		}
-
-		if finished {
-			return
-		}
-	}
+	ctx       context.Context
+	cancel    context.CancelFunc
+	sendCount uint32
+	packet    PRUDPPacketInterface
+	sentAt    time.Time
 }
 
 // ResendScheduler manages the resending of reliable PRUDP packets
 type ResendScheduler struct {
-	packets *MutexMap[uint16, *PendingPacket]
-}
-
-// Stop kills the resend scheduler and stops all pending packets
-func (rs *ResendScheduler) Stop() {
-	stillPending := make([]uint16, rs.packets.Size())
-
-	rs.packets.Each(func(sequenceID uint16, packet *PendingPacket) bool {
-		if !packet.isAcknowledged {
-			stillPending = append(stillPending, sequenceID)
-		}
-
-		return false
-	})
-
-	for _, sequenceID := range stillPending {
-		if pendingPacket, ok := rs.packets.Get(sequenceID); ok {
-			pendingPacket.isAcknowledged = true // * Prevent an edge case where the ticker is already being processed
-
-			if pendingPacket.ticker != nil {
-				// * This should never happen, but popped up in CTGP-7 testing?
-				// * Did the GC clear this before we called it?
-				pendingPacket.ticker.Stop()
-			}
-
-			rs.packets.Delete(sequenceID)
-		}
-	}
+	ctx            context.Context
+	cancel         context.CancelFunc
+	packets        *MutexMap[uint16, *PendingPacket]
+	streamSettings *StreamSettings
 }
 
 // AddPacket adds a packet to the scheduler and begins it's timer
 func (rs *ResendScheduler) AddPacket(packet PRUDPPacketInterface) {
-	connection := packet.Sender().(*PRUDPConnection)
-	slidingWindow := connection.SlidingWindow(packet.SubstreamID())
-
-	pendingPacket := &PendingPacket{
-		packet: packet,
-		rs:     rs,
-		// TODO: This may not be accurate, needs more research
-		interval: time.Duration(slidingWindow.streamSettings.KeepAliveTimeout) * time.Millisecond,
+	sendCount := uint32(1) // This is set to 1 as a packet will already have been sent once when we add it to the scheduler
```

Comment:
This is more clean imo

```suggestion
	var sendCount uint32 = 1 // * This is set to 1 as a packet will already have been sent once when we add it to the scheduler
```

Reply:
I went back and forth on this for ages trying to think what you guys would prefer 😂 I've changed it